### PR TITLE
gossip: tips FIFO cache

### LINKED ISSUE #486: Tips FIFO cache

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #486

# Test Plan:
Added tests

### CODE REVIEW

**Review by th0br0 — APPROVED** [review]

**Review by th0br0 — CHANGES_REQUESTED** [review]
Also add a helper `fifo_pop` method?
Also, could we rename `add` to `push`?

**Review by th0br0 — COMMENTED** [review]
Isn't it counterintuitive to have a FIFO queue backed by a set?

**Comment by thibault-martinez** [comment]
The main motivation for a set was the efficiency of removal.
e.g. removing a tip from a 5000 tips cache.

File: gossip/tips_cache.c [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+  if (hash243_set_size(set) >= capacity) {
+    if ((ret = hash243_set_remove_entry(set, *set)) != RC_OK) {
+      return ret;
+    }
+  }
+
+  return hash243_set_add(set, tip);
+}
+
+static retcode_t tips_cache_random_tip_from_set(hash243_set_t* const set,
```
Comment (th0br0) [inline gossip/tips_cache.c:32]: Move this outside? This is rather a generic `random_hash_from_hash243_set` method?

File: gossip/tips_cache.c [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+retcode_t tips_cache_init(tips_cache_t* const cache, size_t const capacity) {
+  if (cache == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  cache->tips = NULL;
+  cache->solid_tips = NULL;
+  cache->capacity = capacity;
+  rw_lock_handle_init(&cache->lock);
```
Comment (th0br0) [inline gossip/tips_cache.c]: Might be worth adding a note here that we could split this into two distinct locks which might net us a performance improvement

File: gossip/tips_cache.c [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+  if (cache == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_rdlock(&cache->lock);
+
+  if (hash243_set_size(&cache->solid_tips) == 0) {
+    rw_lock_handle_unlock(&cache->lock);
+    return tips_cache_random_tip(cache, tip);
```
Comment (th0br0) [inline gossip/tips_cache.c]: Calling this method here is counterintuitive? This also relocks the rdlock again

File: gossip/tips_cache.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+#ifndef __GOSSIP_TIPS_H__
+#define __GOSSIP_TIPS_H__
+
+#include "common/errors.h"
+#include "common/trinary/flex_trit.h"
+#include "utils/containers/hash/hash243_set.h"
+#include "utils/handles/rw_lock.h"
+
+#define MAX_TIPS 5000
```
Comment (th0br0) [inline gossip/tips_cache.h:16]: not used, remove

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: gossip/tips_cache.h (hunk 4 above)
Comment (thibault-martinez) [reply gossip/tips_cache.h:16]: It's being used on #491

Comment (th0br0) [reply gossip/tips_cache.h:16]: Fair enough; as this is just stored in the `capacity` member I'd much rather not have this in a global header but in the source file that initialises the tips_cache?

Comment (thibault-martinez) [reply gossip/tips_cache.h:16]: Sure. As it can affect the result of `getTips` API call, should it be configurable with an option from CLI ?

Comment (th0br0) [reply gossip/tips_cache.h:16]: :thinking: there's two distinct params here - how many tips are returned from API call and how many are stored in memory. I believe the two are tracked independently on IRI, not sure

(but ya, might be good)

Comment (thibault-martinez) [reply gossip/tips_cache.h:16]: Just looking at IRI conf code, I don't see any option to limit the number of tips returned by the call

File: gossip/tips_cache.c (hunk 2 above)
Comment (thibault-martinez) [reply gossip/tips_cache.c]: I'll add the second lock as part of this PR

File: gossip/tips_cache.h (hunk 4 above)
Comment (thibault-martinez) [reply gossip/tips_cache.h:16]: CLI option will be added in #491

File: gossip/tips_cache.c (hunk 3 above)
Comment (thibault-martinez) [reply gossip/tips_cache.c]: Would you return null hash as well if there is no solid hash ?

**Review by th0br0 — APPROVED** [review]

**Pull request merged**